pkg/applicationserver/applicationserver.go

```diff
@@ -105,7 +105,24 @@ func (as *ApplicationServer) Context() context.Context {
 	return as.ctx
 }
 
-var errListenFrontend = errors.DefineFailedPrecondition("listen_frontend", "failed to start frontend listener `{protocol}` on address `{address}`")
+var (
+	errListenFrontend = errors.DefineFailedPrecondition(
+		"listen_frontend", "failed to start frontend listener `{protocol}` on address `{address}`",
+	)
+	errMaxRetriesReached = errors.DefineAborted(
+		"max_retries_reached", "maximum number of retries ({max_retries}) reached",
+	)
+)
+
+func validateDownlinkConfirmationConfig(c ConfirmationConfig) error {
+	if c.DefaultRetryAttempts == 0 || c.MaxRetryAttempts == 0 {
+		return errZeroAttemptCounts.New()
+	}
+	if c.DefaultRetryAttempts > c.MaxRetryAttempts {
+		return errDefaultRetryCountLargerThanMax.New()
+	}
+	return nil
```

Comment:
> I don't get why `DefaultRetryAttempts` and `MaxRetryAttempts` are both user configurable. Shouldn't the default be hardcoded in the software and checked against a user-set `MaxRetryAttempts`?

Default is used when the user did not provide an explicit value, while max ensures that the user cannot just do an explicit limit of `999999` - we don't want people to retry indefinitely.

Reply:
Sure but my question is this; why is `DefaultRetryAttempts` user configurable?

Reply:
There are 2 configuration options:
1. `DefaultRetryAttempts`, which is used when the user did not provide a maximum amount of retry attempts explicitly in the downlink. User here is the application which submits downlinks.
2. `MaxRettryAttempts`, which is used to validate the user provided maximum amount of retry attempts provided explicitly in a downlink. User here is the application which submits downlinks.

Since `DefaultRetryAttempts` is used when the user did not provide an explicitly limit of retry attempts, to be at least it makes sense to have it configurable. Hardcoding it forces us to re-release the stack in case we want to change the behavior that clients which do not provide an explicit limit have - I don't see why that is preferable.

Reply:
> DefaultRetryAttempts, which is used as a limit when the user did not provide a maximum amount of retry attempts explicitly in the downlink.

Ok why not use `config.MaxRetryAttempts` in this case?

What would be the downside to the following: 
- If user sets the limit explicitly in `ApplicationDownlink`, check that it's not greater than `config.MaxRetryAttempts`
- If user does not set the limit explicitly in `ApplicationDownlink`, use `config.MaxRetryAttempts`. 
And when we read in config, check that this value is not zero.

Reply:
Ok I see, and this ties to your question regarding 'how you came up with these defaults'.

The problem in an abstract is that we lack insight into how many retries somebody needs, and probably that number cannot be universal. Before this change, we retried indefinitely and we didn't have any book keeping on how many attempts did we have so far for a particular downlink. Because of this, we do not really know how many attempts make for a good default.

If we have a singular configuration option, that value may end up being too small (and you cannot request more than the configuration option), or too big (and you retry the same downlink too many times even though it is a lost cause).

If we set the limit too low, the application developers need to add special handling that takes the failed downlink and enqueues it again. If we set it too high, we send the same downlink a lot of times even though it is probably a lost cause.

Having this _default_ in between tries to simplify the transition process between the infinite retries world and the finite retries world. My expectation is that most confirmed downlinks do not need that many retries, so having a smaller default is a sane default. If the user really needs a higher value, the configuration max attempts allows him to provide a higher explicit value without needing to add extra logic that would take the failed downlink and submit it again, manually.

Let me know your thoughts on this.

Reply:
Ok so the problem as I understand, is that you're trying to keep the default no of retransmissions as low as possible while at the same time trying to cap the maximum value to something sane.

This looks good to me then.

An alternative approach would be that the max retries would be a configurable multiple of the default but that's ultimately the same idea.